src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+        cond = %*{},
+        sort = %*{}
+      ): seq[T] {.used.} =
+        let
+          dbCollection = dbConn[customDatabase][getCollectionName(T)]
+        var cursor = dbCollection.find(cond).skip(offset.int32).limit(nLimit.int32)
+        if $sort != $ %*{} :
+          cursor = cursor.orderBy(sort)
+        let fetched = cursor.all()
+        for doc in fetched:
+          var temp = T()
+          applyBSON(temp, doc)
+          result.add temp
+
+      proc pullMany(
+        objs: var seq[object],
+        nLimit: int,
+        offset = 0,
+        cond = %*{},
+        sort = %*{}
+      ) {.used.} =
+        objs = type(objs[^1]).getMany(nLimit, offset, cond, sort)  # this works even on an empty list; which is surprising!
+
+
+      proc update(obj: object, force = false): bool {.used, discardable.} =
+        ##[ Update document with matching ``_id`` with object field values.
+
+        ``force`` parameter ignored.
+
+        returns true if update was successful.
+        ]##
+
+        let
+          dbCollection = dbConn[customDatabase][getCollectionName(type(obj))]
+        var doc = obj.toBson()
+        let response = dbCollection.update(%*{"_id": obj.id}, doc, false, false)
+        if response.n == 1:
+          return true
+        return false
+
+      proc delete(obj: var object): bool {.used, discardable.} =
+        ## Delete a record in DB by object's id. The id is set to all zeroes after the deletion.
+
+        let
+          dbCollection = dbConn[customDatabase][getCollectionName(type(obj))]
+        let response = dbCollection.remove(%*{"_id": obj.id}, 1)
+        obj.id = Oid()
+        if response.n == 1:
+          return true
+        return false
+
+      try:
+        body
+      finally: discard
+
+  template withDb*(body: untyped): untyped {.dirty.} =
+    ##[ A wrapper for actions that require DB connection. Defines CRUD procs to work with the DB,
+    as well as ``createTables`` and ``dropTables`` procs.
+
+      Aforementioned procs and procs defined in a ``db`` block can be used only
+      in  a ``withDb`` block.
+    ]##
+
+    withCustomDb(normConnection, normUser, normPassword, normDatabase):
+      body
+
+
+proc ensureIdFields(typeSection: NimNode): NimNode =
+  ## Check if ``id`` field is in the object definition, insert it if it's not.
+
+  result = newNimNode(nnkTypeSection)
+
+  for typeDef in typeSection:
+    var objRepr = typeDef.toObjRepr()
+
+    if "id" notin objRepr.fieldNames:
+      let idField = FieldRepr(
+        signature: SignatureRepr(
+          name: "id",
+          exported: true,
+          pragmas: @[
+            PragmaRepr(name: "pk", kind: pkFlag),
+            PragmaRepr(name: "ro", kind: pkFlag),
+            PragmaRepr(name: "dbCol", kind: pkKval, value: parseExpr("\"_id\""))
+          ]
+        ),
+        typ: ident "Oid"
+      )
+      objRepr.fields.insert(idField, 0)
+
+    result.add objRepr.toTypeDef()
+
+proc reconstructType(n: NimNode): string =

```

Comment:
Please add docstrings to all procs. In Nim, procs that operate on NimNodes are extremely hard to read, so a doc is a must have.

Reply:
Adding it now.